src/workspace/resolver.rs

```diff
@@ -0,0 +1,373 @@
+
+use std::collections::BTreeMap;
+use std::future::Future;
+
+use anyhow::Error as AnyError;
+use deno_semver::npm::NpmPackageReqReference;
+use deno_semver::package::PackageReqReference;
+use import_map::specifier::SpecifierError;
+use import_map::ImportMap;
+use import_map::ImportMapDiagnostic;
+use import_map::ImportMapError;
+use import_map::ImportMapWithDiagnostics;
+use thiserror::Error;
+use url::Url;
+
+use crate::package_json::PackageJsonDepValueParseError;
+use crate::package_json::PackageJsonDeps;
+use crate::package_json::PackageJsonRc;
+use crate::sync::new_rc;
+use crate::ConfigFile;
+
+use super::UrlRc;
+use super::Workspace;
+
+#[derive(Debug)]
+struct PkgJsonResolverFolderConfig {
+  deps: PackageJsonDeps,
+  pkg_json: PackageJsonRc,
+}
+
+#[derive(Debug, Error)]
+pub enum WorkspaceResolverCreateError {
+  #[error("Failed loading import map specified in '{referrer}'")]
+  ImportMapFetch {
+    referrer: Url,
+    #[source]
+    source: AnyError,
+  },
+  #[error(transparent)]
+  ImportMap(#[from] import_map::ImportMapError),
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct CreateResolverOptions {
+  /// Whether to resolve dependencies by reading the dependencies list
+  /// from a package.json
+  ///
+  /// This should only be set when a node_modules folder doesn't exist
+  /// as node resolution should be used otherwise.
+  pub pkg_json_dep_resolution: bool,
+  pub specified_import_map: Option<SpecifiedImportMap>,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpecifiedImportMap {
+  pub base_url: Url,
+  pub value: serde_json::Value,
+}
+
+#[derive(Debug, Clone)]
+pub enum MappedResolution<'a> {
+  Normal(Url),
+  ImportMap(Url),
+  PackageJson {
+    pkg_json: &'a PackageJsonRc,
+    alias: &'a str,
+    req_ref: NpmPackageReqReference,
+  },
+}
+
+impl<'a> MappedResolution<'a> {
+  pub fn into_url(self) -> Result<Url, url::ParseError> {
+    match self {
+      Self::Normal(url) | Self::ImportMap(url) => Ok(url),
+      Self::PackageJson { req_ref, .. } => Url::parse(&req_ref.to_string()),
+    }
+  }
+}
+
+#[derive(Debug, Error)]
+pub enum MappedResolutionError {
+  #[error(transparent)]
+  Specifier(#[from] SpecifierError),
+  #[error(transparent)]
+  ImportMap(#[from] ImportMapError),
+  #[error(transparent)]
+  PkgJsonDep(#[from] PackageJsonDepValueParseError),
+}
+
+impl MappedResolutionError {
+  pub fn is_unmapped_bare_specifier(&self) -> bool {
+    match self {
+      MappedResolutionError::Specifier(err) => match err {
+        SpecifierError::InvalidUrl(_) => false,
+        SpecifierError::ImportPrefixMissing { .. } => true,
+      },
+      MappedResolutionError::ImportMap(err) => match err {
+        ImportMapError::UnmappedBareSpecifier(_, _) => true,
+        ImportMapError::Other(_) => false,
+      },
+      MappedResolutionError::PkgJsonDep(_) => false,
+    }
+  }
+}
+
+#[derive(Debug)]
+pub struct WorkspaceResolver {
+  maybe_import_map: Option<ImportMapWithDiagnostics>,
```

Comment:
I ended up making this just use a single import map like we currently do because it was a lot simpler and was the desired behaviour from the RFC.

Reply:
What's the result of that? We no longer create "synthetic" source maps?

Reply:
No, this is the same as before. We were discussing not creating a single import map.